File: src/test/java/io/github/petrknap/persistence/zoneddatetime/UtcWithSystemTimezoneTest.java
package io.github.petrknap.persistence.zoneddatetime;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.Test;

import java.time.LocalDateTime;
import java.time.ZonedDateTime;

import static org.junit.jupiter.api.Assertions.*;

final class UtcWithSystemTimezoneTest extends UtcTestCase
{
    @Override @Test void constructs_itself()
    {
        assertInstance(
                getInstance(zonedDateTime),
                utcDateTime.toLocalDateTime()
        );
    }

    @Override @Test void fromValues()
    {
        assertInstance(
                UtcWithSystemTimezone.fromValue(
                        utcDateTime.toLocalDateTime()
                ),
                utcDateTime.toLocalDateTime()
        );
    }

    @Override @Test void fromValues_of_null()
    {
        assertNull(UtcWithSystemTimezone.fromValue(null));
    }

    @Override @Test void fromFormattedValues()
    {
        assertInstance(
                UtcWithSystemTimezone.fromFormattedValue(
                        localDateTimeFormatter.format(utcDateTime),
                        LOCAL_DATETIME_FORMAT
                ),
                utcDateTime.toLocalDateTime()
        );
    }

    @Override @Test void fromFormattedValues_of_null()
    {
        assertNull(UtcWithSystemTimezone.fromFormattedValue(null, LOCAL_DATETIME_FORMAT));
    }

    @Override protected @NotNull UtcWithSystemTimezone getInstance(@NotNull ZonedDateTime zonedDateTime)
    {
        return new UtcWithSystemTimezone(zonedDateTime);
    }

    private static void assertInstance(
            @Nullable UtcWithSystemTimezone actual,
            @NotNull LocalDateTime expectedUtcDateTime
    ) {
        assertNotNull(actual);
        assertEquals(
                expectedUtcDateTime,
                actual.getUtcDateTime(),
                "Unexpected UTC date-time"
        );
    }
}

File: src/test/php/UtcWithSystemTimezoneTest.php
<?php

declare(strict_types=1);

namespace PetrKnap\Persistence\ZonedDateTime;

use DateTimeImmutable;

/**
 * @phpstan-import-type LocalDateTime from JavaSe8\Time
 */
final class UtcWithSystemTimezoneTest extends UtcTestCase
{
    public function test_constructs_itself(): void
    {
        self::assertInstance(
            $this->getInstance($this->zonedDateTime),
            JavaSe8\Time::toLocalDateTime($this->utcDateTime),
        );
    }

    public function test_fromValues(): void
    {
        self::assertInstance(
            UtcWithSystemTimezone::fromValue(
                $this->utcDateTime,
            ),
            JavaSe8\Time::toLocalDateTime($this->utcDateTime),
        );
    }

    public function test_fromValues_of_null(): void
    {
        self::assertNull(UtcWithSystemTimezone::fromValue(null));
    }

    public function test_fromFormattedValues(): void
    {
        self::assertInstance(
            UtcWithSystemTimezone::fromFormattedValue(
                $this->utcDateTime->format(self::LOCAL_DATETIME_FORMAT),
                self::LOCAL_DATETIME_FORMAT,
            ),
            JavaSe8\Time::toLocalDateTime($this->utcDateTime),
        );
    }

    public function test_fromFormattedValues_of_null(): void
    {
        self::assertNull(UtcWithSystemTimezone::fromFormattedValue(null, self::LOCAL_DATETIME_FORMAT));
    }

    protected function getInstance(DateTimeImmutable $zonedDateTime): UtcWithSystemTimezone
    {
        return new UtcWithSystemTimezone($zonedDateTime);
    }

    /**
     * @param LocalDateTime $expectedUtcDateTime
     */
    private static function assertInstance(
        UtcWithSystemTimezone|null $actual,
        DateTimeImmutable $expectedUtcDateTime,
    ) {
        self::assertNotNull($actual);
        self::assertEquals(
            $expectedUtcDateTime,
            $actual->getUtcDateTime(),
            'Unexpected UTC date-time'
        );
    }
}
